tests/index-test.ts

```diff
@@ -207,6 +207,54 @@ describe('TSDI', () => {
       assert.isTrue(called);
     });
 
+    it('should call the initializer of an async dependency with a factory', done => {
+      tsdi.enableComponentScanner();
+
+      class Instanced {
+        public value?: number;
+
+        constructor(value: number) {
+          this.value = value;
+        }
+      }
+
+      // @ts-ignore
+      @component
+      // @ts-ignore
+      class AsyncFactory {
+        private instance!: Instanced;
+
+        @initialize
+        protected async init(): Promise<void> {
+          return new Promise<void>(resolve => {
+            setTimeout(() => {
+              this.instance = new Instanced(10);
+              resolve();
+            }, 10);
+          });
+        }
+
+        @factory
+        public getInstance(): Instanced {
+          return this.instance;
+        }
+      }
+
+      @component
+      class Dependent {
+        @inject public instance!: Instanced;
+
+        @initialize
+        protected init(): void {
+          assert.equal(this.instance.value, 10);
+          done();
+        }
+      }
+
+      // todo: this must throw because Dependent is async
+      tsdi.get(Dependent);
```

Comment:
@Prior99  If we let this throw, it means async initializers could only be used during injections, but not to 'kick-off' a dependency chain. Usally when you want to start an application with TSDI you have a `tsdi.get(Dependent)` somewhere. If thats async you are not able to start your app.
I would suggest to return a Promise here if the dependency is async.
Any thoughts?

Reply:
This is implemented in https://github.com/KnisterPeter/tsdi/pull/1083

Reply:
To be honest I didn't add this, I think it was left over from your changes. I wouldn't touch the interface of `tsdi.get`. Async dependencies are available immediately, as as you already mentioned, the constructor shouldn't return a promise. Hence the `tsdi.get` can always return immediately.

The only problem is that the `@initialize` hook is called too early.

I would expect the following order of initialization:

1. Instances of components and externals are constructed synchronously 
2. All initialize hooks that have no asynchronous dependencies are called synchronously
3. All initalize hooks that have asynchronous dependencies are called in the background. 

Perhaps a `@asyncInitialize` and a `tsdi.getAsync` would separate step 3 from step 2.
However, I find it rather confusing that `@initialize` doesn't wait for all dependencies to be ready. I would expect all dependencies to be there and usable by the time `@initialize` is called. Would an optional parameter in `@initialize` perhaps solve this? (`@initialize({ waitForAsync: true}) `.